framework/test/gtest/BundleTrackerConcurrencyTest.cpp

```diff
@@ -0,0 +1,232 @@
+              (const Bundle&, const BundleEvent&, const Bundle&),
+              (override));
+};
+
+TEST_F(BundleTrackerConcurrencyTest, ConcurrentOpenCloseWorks)
+{
+  auto bundleTracker = std::make_unique<BundleTracker<>>(context, all_states);
+
+  size_t numThreads = std::thread::hardware_concurrency();
+  ASSERT_GT(numThreads, 0ull) << "number of threads is 0";
+  std::vector<std::future<void>> futures;
+  std::promise<void> gate;
+  auto gateFuture = gate.get_future().share();
+  for (size_t i = 0; i <= numThreads; ++i) {
+    futures.push_back(
+      std::async(std::launch::async, [i, &bundleTracker, &gateFuture]() {
+        gateFuture.get();
+        for (int n = 0; n < 1000; ++n) {
+          if (i % 2 == 0) {
+            bundleTracker->Open();
+          } else {
+            bundleTracker->Close();
+          }
+        }
+      }));
+  }
+
+  gate.set_value();
+
+  for (auto& asyncFuture : futures) {
+    asyncFuture.get();
+  }
+}
+
+TEST_F(BundleTrackerConcurrencyTest, OpeningTrackerWhileBundlesChangeWorks)
+{
+  auto customizer = std::make_shared<MockCustomizer2>();
+  auto stateMask =
+    BundleTracker<>::CreateStateMask(Bundle::State::STATE_ACTIVE);
+  auto bundleTracker =
+    std::make_unique<BundleTracker<>>(context, stateMask, customizer);
+
+  size_t numThreads = std::thread::hardware_concurrency();
+  ASSERT_GT(numThreads, 0ull) << "number of threads is 0";
+  std::vector<std::future<void>> futures;
+  std::promise<void> gate;
+  auto gateFuture = gate.get_future().share();
+
+  auto openTrackerFuture =
+    std::async(std::launch::async, [&bundleTracker, &gateFuture]() {
+      gateFuture.get();
+      bundleTracker->Open();
+    });
+
+  auto startBundlesFuture =
+    std::async(std::launch::async, [this, &gateFuture]() {
+      gateFuture.get();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleA")
+        .Start();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleA2")
+        .Start();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleB")
+        .Start();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleH")
+        .Start();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleM")
+        .Start();
+      cppmicroservices::testing::InstallLib(framework.GetBundleContext(),
+                                            "TestBundleR")
+        .Start();
+    });
+
+  EXPECT_CALL(*customizer, AddingBundle).Times(7);
+  gate.set_value();
+
+  openTrackerFuture.get();
+  startBundlesFuture.get();
+
+  bundleTracker->Close();
+}
+
+TEST_F(BundleTrackerConcurrencyTest, NoRaceConditionForRemovingChangingBundle)
```

Comment:
minor; rename testpoint name - prepend "Test" to be more consistent with other test point names

Reply:
Done.